### PR TITLE
Add BeforePackEvent event call prior to packing

### PR DESCRIPTION
**Pull request by gig745**
Add BeforePackEvent event call prior to packing to allow modules to select  which order items that should be packed.

### CODE REVIEW

File: src/Event/CommerceAuspostEvents.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace Drupal\commerce_auspost\Event;
+
+class CommerceAuspostEvents {
```
Comment (chinthakagodawita) [inline src/Event/CommerceAuspostEvents.php]: Needs a class comment.

Could you also make this a `final` class and extend `Enum`?

File: src/Plugin/Commerce/ShippingMethod/AusPost.php [hunk 2]
```diff
@@ -227,13 +237,15 @@ public function __construct(
     ConfigureForm $configurationForm,
     ServiceSupport $serviceSupport,
     ServiceDefinitionManager $serviceManager,
-    ClientInterface $client
+    ClientInterface $client,
+    EventDispatcherInterface $eventDispatcher
```
Comment (chinthakagodawita) [inline src/Plugin/Commerce/ShippingMethod/AusPost.php:243]: `$eventDispatcher` needs to be added to the constructor's docblock.

File: src/Plugin/Commerce/ShippingMethod/AusPost.php [hunk 3]
```diff
@@ -399,9 +412,18 @@ public function calculateRates(ShipmentInterface $shipment) {
       $postagePrice = new Price(0, static::AUD_CURRENCY_CODE);
 
       try {
+        $event = $this->eventDispatcher->dispatch(
+          CommerceAuspostEvents::BEFORE_PACK,
+          new BeforePackEvent($shipment->getOrder()->getItems())
+        );
+
+        if (count($event->getOrderItems()) === 0) {
+          return [];
```
Comment (chinthakagodawita) [inline src/Plugin/Commerce/ShippingMethod/AusPost.php]: Can you change this to return `$rates` instead?

**Review by chinthakagodawita — CHANGES_REQUESTED** [review]
@jrhodes-sitback Just needs a few changes.

File: src/Plugin/Commerce/ShippingMethod/AusPost.php (hunk 2 above)
Comment (gig745) [reply src/Plugin/Commerce/ShippingMethod/AusPost.php:243]: Fixed

File: src/Event/CommerceAuspostEvents.php (hunk 1 above)
Comment (gig745) [reply src/Event/CommerceAuspostEvents.php]: Fixed

**Review by chinthakagodawita — APPROVED** [review]

**Comment by chinthakagodawita** [comment]
Looks good to me 👍

**Pull request merged**